lib/patchelf/alt_saver.rb

```diff
@@ -0,0 +1,920 @@
+    def dyn_tag_to_section_name(d_tag)
+      case d_tag
+      when ELFTools::Constants::DT_STRTAB, ELFTools::Constants::DT_STRSZ
+        '.dynstr'
+      when ELFTools::Constants::DT_SYMTAB
+        '.dynsym'
+      when ELFTools::Constants::DT_HASH
+        '.hash'
+      when ELFTools::Constants::DT_GNU_HASH
+        '.gnu.hash'
+      when ELFTools::Constants::DT_JMPREL
+        sec_name = %w[.rel.plt .rela.plt .rela.IA_64.pltoff].find { |s| find_section(s) }
+        raise PatchError, 'cannot find section corresponding to DT_JMPREL' unless sec_name
+
+        sec_name
+      when ELFTools::Constants::DT_REL
+        # regarding .rel.got, NixOS/patchelf says
+        # "no idea if this makes sense, but it was needed for some program"
+        #
+        # return nil if not found, patchelf claims no problem in skipping
+        %w[.rel.dyn .rel.got].find { |s| find_section(s) }
+      when ELFTools::Constants::DT_RELA
+        # return nil if not found, patchelf claims no problem in skipping
+        find_section('.rela.dyn')&.name
+      when ELFTools::Constants::DT_VERNEED
+        '.gnu.version_r'
+      when ELFTools::Constants::DT_VERSYM
+        '.gnu.version'
+      end
+    end
+
+    # updates dyn tags by syncing it with @section values
+    def sync_dyn_tags!
+      each_dynamic_tags do |dyn, buf_off|
+        sec_name = dyn_tag_to_section_name(dyn.d_tag)
+        next unless sec_name
+
+        shdr = find_section(sec_name).header
+        dyn.d_val = dyn.d_tag == ELFTools::Constants::DT_STRSZ ? shdr.sh_size.to_i : shdr.sh_addr.to_i
+
+        with_buf_at(buf_off) { |wbuf| dyn.write(wbuf) }
+      end
+    end
+
+    def update_section_idx!
+      @section_idx_by_name = @sections.map.with_index { |sec, idx| [sec.name, idx] }.to_h
+    end
+
+    def with_buf_at(pos)
+      return unless block_given?
+
+      opos = @buffer.tell
+      @buffer.seek pos
+      yield @buffer
+      @buffer.seek opos
+      nil
+    end
+
+    def sync_sec_to_seg(shdr, phdr)
+      phdr.p_offset = shdr.sh_offset.to_i
+      phdr.p_vaddr = phdr.p_paddr = shdr.sh_addr.to_i
+      phdr.p_filesz = phdr.p_memsz = shdr.sh_size.to_i
+    end
+
+    def phdrs_by_type(seg_type)
+      return unless seg_type
+
+      @segments.each_with_index do |seg, idx|
+        next unless (phdr = seg.header).p_type == seg_type
+
+        yield phdr, idx
+      end
+    end
+
+    def find_matching_note_segment_idx(orig_shdr, skip_indices: nil)
+      phdrs_by_type(ELFTools::Constants::PT_NOTE) do |phdr, seg_idx|
+        next if skip_indices.member?(seg_idx)
+
+        sec_range = (orig_shdr.sh_offset.to_i)...(orig_shdr.sh_offset + orig_shdr.sh_size)
+        seg_range = (phdr.p_offset.to_i)...(phdr.p_offset + phdr.p_filesz)
+        next unless seg_range.cover?(sec_range.first) || seg_range.cover?(*sec_range.last(1))
+
+        raise PatchError, 'unsupported overlap of SHT_NOTE and PT_NOTE' if seg_range != sec_range
+
+        return seg_idx
+      end
+    end
```

Comment:
`nil` is needed before this `end`, otherwise the return value will be the one returned by `phdrs_by_type`, which is an array of indexes